[jit] Added a pass to transform aten::cat ops to prim::Concat op with variable number of inputs

navahgar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59881 [jit] Added a pass to transform aten::cat ops to prim::Concat op with variable number of inputs**
* #59880 [jit] Added a new operator for concat that takes in variadic parameters

This pass is not included in the JIT flow or anywhere else at this point. The idea is, once this lands, everyone can use this to test their workflow with this transformation and once we are convinced this is useful and/or improves performance, we can include it in the appropriate workflow.

Differential Revision: [D29277876](https://our.internmc.facebook.com/intern/diff/D29277876)

navahgar committed:
[jit] Added a pass to transform aten::cat ops to prim::Concat op with variable number of inputs

[ghstack-poisoned]

navahgar committed:
Update on "[jit] Added a pass to transform aten::cat ops to prim::Concat op with variable number of inputs"




[ghstack-poisoned]

eellison commented on torch/csrc/jit/passes/concat_opt.cpp:
```diff
@@ -510,5 +510,65 @@ void OptimizeConcat(const std::shared_ptr<Graph>& graph) {
+    // because these uses could be modifying the list using ops like
+    // `aten::append`. So, we conservatively assume that any use other than
+    // the one in cat mutates the list.
+    if (list->output()->uses().size() > 1) {
```
Just doing a quick review, I think we want this pass to work by testing the `aliasDb. moveBeforeTopologicallyValid`, that way we can get the concat optimization pass you wrote firing with DensetNet even though the list is appended to. What do you think ?

navahgar replied:
As we discussed in the meeting, the plan would be to remove list mutation before doing this transformation. Given that, why do we need to use `aliasDb.moveBeforeTopologicallyValid` here? 

I definitely want to apply the concat optimization pass for DenseNet, but that can be done as soon as we eliminate list mutation.

penguinwu commented on test/cpp/jit/test_concat_opt.cpp:
```diff
@@ -600,5 +600,129 @@ TEST(OptimizeConcatTest, NoOptimizationWhenInputListIsMutated) {
+      ->run(*graph);
+}
+
+TEST(OptimizeConcatTest, UseVariadicCat2) {
```
The second test case looks just like the first one except for more elements in the list. Are they exercising different logics in the implementation? If not, perhaps keeping UseVariadicCat2 would be sufficient.

Maybe we should add a test case where the list constructed is not only used in concat but also used in another computation. The test will check that the constructed list is not preserved.

```
    R"IR(        graph(%0: Float(64, 56, 56, strides=[3136, 56, 1], requires_grad=0, device=cpu), 
             %1: Float(32, 56, 56, strides=[3136, 56, 1], requires_grad=0, device=cpu)): 
             %2 : int = prim::Constant[value=0]()
             %input : Tensor[] = prim::ListConstruct(%0, %1) 
            %concat : Float(96, 56, 56, strides=[3136, 56, 1], requires_grad=0, device=cpu) = aten::cat(%input, %2)
            return (%concat,%input)      )IR";
```

navahgar replied:
Sure, added that test case. Retained only one for the positive case.

Note that, with the case of List output being used in multiple places, the transformation would not happen now, since we are explicitly checking for > 1 use of the List.

penguinwu commented on test/cpp/jit/test_concat_opt.cpp:
```diff
@@ -600,5 +600,129 @@ TEST(OptimizeConcatTest, NoOptimizationWhenInputListIsMutated) {
+  auto opt_outputs = runGraph(graph, inputs);
+  checkOutputs(orig_outputs, opt_outputs);
+
+  // No transformation should have happened since the `prim::ListConstruct` has
```
If you run mutation removal first, I think it will get rid of `at::apend`, then you should be able to generate `prim::Cat`. Please give it a try.

navahgar replied:
I plan to do that separately with a higher level API. Actually the mutation removal and the transformation to use variadic cat should be performed multiple times, until a fixed point.

I want to keep this API just for the basic transformation of replacing with variadic cat.

penguinwu commented on torch/csrc/jit/passes/concat_opt.cpp:
```diff
@@ -510,5 +510,65 @@ void OptimizeConcat(const std::shared_ptr<Graph>& graph) {
+  explicit VariadicCatUpdater(std::shared_ptr<Graph> graph)
+      : graph_(std::move(graph)) {}
+
+  void run() {
```
You may want to run mutation removal pass first because that pass may remove some list mutation operations

navahgar replied:
Same reason as above. Will add another higher level API to do both these until a fixed point.

penguinwu commented on torch/csrc/jit/passes/concat_opt.cpp:
```diff
@@ -510,5 +510,65 @@ void OptimizeConcat(const std::shared_ptr<Graph>& graph) {
+
+void UseVariadicCat(const std::shared_ptr<Graph>& graph) {
+  GRAPH_DUMP("Before VariadicCat", graph);
+  VariadicCatUpdater(graph).run();
```
To follow the convention of other passes, make the transformation return a boolean to indicate whether graph is modified and print the graph only if graph is transformed.

navahgar replied:
Sure, good point. Done.

navahgar committed:
Update on "[jit] Added a pass to transform aten::cat ops to prim::Concat op with variable number of inputs"


This pass is not included in the JIT flow or anywhere else at this point. The idea is, once this lands, everyone can use this to test their workflow with this transformation and once we are convinced this is useful and/or improves performance, we can include it in the appropriate workflow.

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29277876).

navahgar committed:
Update on "[jit] Added a pass to transform aten::cat ops to prim::Concat op with variable number of inputs"


This pass is not included in the JIT flow or anywhere else at this point. The idea is, once this lands, everyone can use this to test their workflow with this transformation and once we are convinced this is useful and/or improves performance, we can include it in the appropriate workflow.

Differential Revision: [D29277876](https://our.internmc.facebook.com/intern/diff/D29277876)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29277876).

navahgar committed:
Update on "[jit] Added a pass to transform aten::cat ops to prim::Concat op with variable number of inputs"


This pass is not included in the JIT flow or anywhere else at this point. The idea is, once this lands, everyone can use this to test their workflow with this transformation and once we are convinced this is useful and/or improves performance, we can include it in the appropriate workflow.

Differential Revision: [D29277876](https://our.internmc.facebook.com/intern/diff/D29277876)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29277876).

navahgar committed:
Update on "[jit] Added a pass to transform aten::cat ops to prim::Concat op with variable number of inputs"


This pass is not included in the JIT flow or anywhere else at this point. The idea is, once this lands, everyone can use this to test their workflow with this transformation and once we are convinced this is useful and/or improves performance, we can include it in the appropriate workflow.

Differential Revision: [D29277876](https://our.internmc.facebook.com/intern/diff/D29277876)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29277876).
